src/sdk/account/utils/Utils.ts

```
@@ -620,3 +629,106 @@ export function calculateNonceStorageSlot(sender: Address, key = 0n): Hex {
+ * signature validation across chains.
+ *
+ * @param meeVersions - Array of chain IDs with their corresponding MEE versions
+ * @throws Error if versions are inconsistent or mixing signing methods
+ */
+export function validateConsistentMeeVersions(
+  meeVersions: MeeVersionsWithChainId
+): void {
+  let hasPersonalSignVersion = false
+  let hasEIP712Version = false
+  const meeVersionStrings = new Set<string>()
+
+  // Track version consistency while iterating (fail-fast approach)
+  for (const { version: meeVersionConfig } of meeVersions) {
+    const meeVersion = meeVersionConfig.version
+    // Check version type as we go
+    const isCurrentVersionPersonalSign = isVersionOlder(
+      meeVersion,
+      MEEVersion.V2_2_0
+    )
+    const isCurrentVersionEIP712 = versionMeetsRequirement(
+      meeVersion,
+      MEEVersion.V2_2_0
+    )
+
+    // Update flags
+    hasPersonalSignVersion =
+      hasPersonalSignVersion || isCurrentVersionPersonalSign
+    hasEIP712Version = hasEIP712Version || isCurrentVersionEIP712
+
+    // Early exit: Cannot mix personal_sign and EIP-712 signing methods
+    if (hasPersonalSignVersion && hasEIP712Version) {
+      throw new Error(
+        "MEE versions on all chains should be whether less than 2.2.0 or greater than or equal to 2.2.0. " +
+          "Otherwise Multichain account won't be able to consume the same signature across all chains involved in the quote request."
+      )
+    }
+
+    // Only collect version strings if we're using EIP-712
+    if (isCurrentVersionEIP712) {
+      meeVersionStrings.add(meeVersion)
+
+      // Early exit: if we have more than one MEE version fail immediately
+      if (meeVersionStrings.size > 1) {
+        throw new Error(
+          "All the MEE versions on all chains are greater than or equal to 2.2.0 in your quote request. " +
+            "In this case, MEE versions should be same for all chains within quote request."
+        )
```

So why are we restricting more than one version if all these different versions are EIP712 compatible.

How I see this is,
All the personal sign version can be grouped and mixed with different versions and same goes for all the versions with 712.

But if personal sign and 712 are mixed ? We can throw an error

---

Because eip-712 versions involve domain separator which includes version. 
To me, it makes sense keeping version in the domain sep
Since when we do eip-712, there are more potential changes and parameters, it can be that some >= 2.2.0 versions are also incompatible

---

I am concerned about this scenario for future similar to transient storage and 7702 stuff we did.

Where we may need to deploy new nexus on some chains and some doesn't support new EVM features. In these cases, the supertransaction cannot include all the chains as before with different MEE versions

---

We cannot force them to use same version for all chains in these cases

---

Gonna rm this check
so can mix < 2.2.0 and >= 2.2.0